crates/ruff_db/src/vendored.rs

```diff
@@ -240,45 +241,86 @@ impl VendoredZipArchive {
-            camino::Utf8Component::Normal(part) => normalized_parts.push(part),
-            camino::Utf8Component::CurDir => continue,
-            camino::Utf8Component::ParentDir => {
-                normalized_parts.pop();
+impl<'a> From<&'a VendoredPath> for NormalizedVendoredPath<'a> {
+    /// Normalizes the path by removing `.` and `..` components.
+    ///
+    /// ## Panics:
+    /// If a path with an unsupported component for vendored paths is passed.
+    /// Unsupported components are path prefixes and path root directories.
+    fn from(path: &'a VendoredPath) -> Self {
+        // Attempt to avoid allocating unless the path is unnormalized in some way
+        // (in most cases, it should hopefully already be normalized)
+
+        let mut normalized_parts: Option<Vec<&str>> = {
+            let path_str = path.as_str();
+            // If the path uses Windows separators or has a trailing slash,
+            // it requires normalization, so there's no way to avoid allocating;
+            // may as well do so upfront
+            if (std::path::MAIN_SEPARATOR == '\\' && path_str.contains('\\'))
+                || path_str.ends_with('/')
+            {
+                Some(Vec::new())
+            } else {
+                None
+            }
+        };
+
+        for (i, component) in path.components().enumerate() {
+            match component {
+                camino::Utf8Component::Normal(part) => {
+                    if let Some(ref mut data) = normalized_parts {
+                        data.push(part);
+                    }
+                }
+                camino::Utf8Component::CurDir => {
+                    if normalized_parts.is_none() {
+                        normalized_parts = Some(
+                            path.components()
+                                .take(i)
+                                .map(|part| part.as_str())
+                                .collect(),
+                        );
+                    }
+                }
+                camino::Utf8Component::ParentDir => {
+                    if let Some(ref mut data) = normalized_parts {
+                        data.pop();
+                    } else {
+                        normalized_parts = Some(
+                            path.components()
+                                .take(i - 1)
+                                .map(|part| part.as_str())
+                                .collect(),
+                        );
+                    }
+                }
+                unsupported => panic!("Unsupported component in a vendored path: {unsupported}"),
             }
-            unsupported => panic!("Unsupported component in a vendored path: {unsupported}"),
         }
+
+        NormalizedVendoredPath({
+            if let Some(normalized) = normalized_parts {
+                Cow::Owned(normalized.join("/"))
+            } else {
+                Cow::Borrowed(path.as_str())
+            }
+        })
```

Comment:
I think we can simplify the implementation and also reduce the branching by doing a simple search first:

1. Test if the path contains any `\`, if so, jump right into the normalization logic
2. Test if all components are `Normal`. If so, return `Cow::Borrowed(path.as_str())`
3. Otherwise run the normlaization